Wrote a UMAPVisualizer for text

jc-healy opened the pull request:
After seeing Rebecca speak at PyDataNY I promised her a text/UMAPVisualizer as a drop in replacement for the current text/TSNEVisualizer currently in Yellowbrick.  UMAP has a few advantages over tSNE in the case of text.  One of the primary ones is that it can run directly on the sparse token vectors generated by a CountVectorizer and thus doesn't require an SVD or PCA to be run on the sparse matrix as a pre-processing step.  A few other nice practical wins are that it is quite fast and tends to preserve more of the global structure within an embedding.  The better preservation of global structure makes the distances between clusters more meaningful which can be useful for interpretability.  

I've added documentation to your docs/api/text directory and a integrated some unit tests as well.

Let me know if you need me to alter or fix anything else and I'll be happy to.

jc-healy commented:
Hmm... I'm trying to make Travis happy but I'm not sure what's going on.  When I run the make test locally I get no errors.  When Travis runs the unit tests it claims that images generated don't match those in the reference directory.  Any idea why that might differ from my local box?  Here are the test results from a clean environment locally:

UMAPVisualizer tests (tests/test_text/test_umap.py)
    [PASS]  Test that the user can supply an alpha param on instantiation
    [PASS]  Check UMAP can accept a custom size (tuple of pixels) from the user
    [PASS]  Check UMAP can accept a custom title (string) from the user
    [PASS]  Check UMAP integrated visualization on the hobbies corpus
    [PASS]  Test UMAP integrated visualization on a sklearn classifier dataset
    [PASS]  Test UMAP integrated visualization with class labels specified
    [PASS]  Verify the pipeline creation step for UMAP
    [PASS]  Test UMAP when no target or classes are specified
    [PASS]  Check to make sure sklearn's UMAP doesn't use the size param
    [PASS]  Check to make sure sklearn's UMAP doesn't use the title param
    [PASS]  Assert exception is raised when number of labels doesn't match
    [PASS]  Test UMAP when passed a pandas DataFrame and series

lwgray commented:
@jc-healy Thank you so much for this wonderful PR.  We will try to get this reviewed on Monday.  

The errors that you are receiving could be related to tolerance levels in "assert_images_similar". You should first try to increase the tolerance of each.  The errors can also be related to the OS you are using.  Can you tell me what OS you are working on?

bbengfort commented:
@jc-healy if you're generating them on Linux they should be close to what's happening on Travis, so, unfortunately, I'm not sure what's happening. Maybe centos has a different font? I think updating the tolerances is the right move for now; once you've made your changes I can also run them on my machine to see if that makes a difference. 

Thanks again for contributing this visualizer!

jc-healy commented on yellowbrick/text/umap_vis.py:
```diff
@@ -0,0 +1,351 @@
+        # Call super at the end so that size and title are set correctly
+        super(UMAPVisualizer, self).__init__(ax=ax, **kwargs)
+
+    def make_transformer(self, umap_kwargs={}):
```
I have a preference for leaving it in if you are OK with that.  It provides some nice consistency between the two classes and makes the code work as a nice template in case one wants to extend or modify the classes with any extra transformations at a later date.

bbengfort replied:
That's good reasoning and I'm willing to go with it - we'll keep it in!

jc-healy commented:
Ok, I removed the conditional statements from the unit tests.  The default tolerances were sufficiently high that all tests pass and Travis is currently happy.  

My next problem is with appveyor.  UMAP is depends on the wonderful numba for it's performance and unfortunately numba doesn't support python 2.7 for 32 bit windows.  As such appveyor fails on this case.  I've added an informative RuntimeError and warning to  umap_vis.py which is being triggered but I'm not sure what the next step would be in dealing with this problem.  Any advice for dealing with an unsupported OS/python version?

bbengfort commented:
Hi @jc-healy - thanks for making the changes I requested, looking forward to getting this merged in! 

For the AppVeyor tests, we usually do something like:

```python
import sys
import pytest 

@pytest.mark.xfail(
    sys.platform == 'win32', reason="not supported on windows 32bit with Python 2.7"
)
```

Raising the `RuntimeError` and warning is also appreciated, thank you for being so thoughtful! 

Note also that in Yellowbrick 1.0 we're removing support for Python 2 (following numpy and scipy) -- so this may not be a problem for too much longer. Do the tests pass on Windows with Python 3?

jc-healy commented:
Thanks for the xfail example.  I've decorated all the test_umap.py functions with the win32 check and caught the AttributeError that was being thrown on import of umap under win32.  All test seem to be good to go now.  I tried to limit the xfail cases to wind32 and python 2.7 via this:
`@pytest.mark.xfail((sys.platform == 'win32') and (sys.version[0] < 3), reason="not supported on windows 32bit with Python 2.7")`

Unfortunately, that didn't seem to work.  I'm happy to try something else but I'm not much of a pytest guru.  I believe that it should run for python 3.6 on win32 but haven't been able to test it because I haven't been able to get past the win32 python 2.7 test case without xfailing all of win32.

I've reverted to:
`@pytest.mark.xfail(sys.platform == 'win32', reason="not supported on windows 32bit with Python 2.7")`
which works.

Glad to hear that yellowbrick 1.0 is moving away from Python 2 support.  That will likely make things easier going forward.

jc-healy commented:
Hi @bbengfort, I think I've got everything looking reasonably good on my end of things.  Let me know if you need me to take care of any other issues before folding this in.

bbengfort commented:
@jc-healy thank you so much! Sorry for the delay; I've been traveling with family for the holiday. I'll take a look at this soon - ping me if you don't hear from me by next week!
